### PR TITLE
Fix some BSD builds missing pthread functions

### PR DESCRIPTION
**Pull request by Ericson2314**
# Motivation

In addition to adding the missing thread deps in the last commit, we also appear to need to skip `-Wl,--as-needed` flags that Meson wants to use, but doesn't work with our *BSD toolchains.

Also, add missing thread deps.

# Context

See https://github.com/mesonbuild/meson/issues/3593

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

**Review by tomberek — APPROVED** [review]
Builds as expected.

File: packaging/dependencies.nix [hunk 1]
```diff
@@ -61,6 +61,14 @@ let
       workDir = null;
     };
 
+  # Work around weird `--as-needed` linker behavior with BSD, see
+  # https://github.com/mesonbuild/meson/issues/3593
+  bsdNoLinkAsNeeded = finalAttrs: prevAttrs: lib.optionalAttrs stdenv.hostPlatform.isBSD {
+    postConfigure = prevAttrs.postConfigure or "" + ''
+       sed -i build.ninja -e 's/-Wl,--as-needed //'
+    '';
+  };
```
Comment (eli-schwartz) [inline packaging/dependencies.nix:70]: Please instead use the meson option `-D b_asneeded=false` to disable generating this option at all.

**Review by eli-schwartz — CHANGES_REQUESTED** [review]

File: packaging/dependencies.nix (hunk 1 above)
Comment (Ericson2314) [reply packaging/dependencies.nix]: Good idea, switched it!

**Comment by edolstra** [comment]
@Ericson2314 There still appears to be a pthread-related issue on FreeBSD: https://hydra.nixos.org/build/267279816

Note that the need for `pthread_getattr_np()` would go away with #11152 so it might not be worth investing too much time in fixing it.

**Pull request merged**